posts/2017-05-02-io-monad-for-cats.md

```diff
@@ -0,0 +1,404 @@
+
+For example, here's a program that performs some simple user interaction in the shell:
+
+```tut:silent
+import cats.effect.IO
+
+val program = for {
+  _ <- IO { println("Welcome to Scala!  What's your name?") }
+  name <- IO { Console.readLine }
+  _ <- IO { println(s"Well hello, $name!") }
+} yield ()
+```
+
+We could have just as easily written this program in the following way:
+
+```tut:silent
+val program = IO {
+  println("Welcome to Scala!  What's your name?")
+  val name = Console.readLine
+  println(s"Well hello, $name!")
+}
+```
+
+But this gives us less flexibility for composition.  Remember that even though `program` is a pure and referentially transparent value, its *definition* is not.  Anything inside the `IO {` … `}` block is not referentially transparent, and so should be treated with extreme care and suspicion.  The less of our program we have inside these blocks, the better!
+
+As a sidebar that is actually kinda cool, we can implement a `readString` `IO` action that wraps `Console.readLine` *as a `val`!*
+
+```tut:silent
+val readString = IO { Console.readLine }
+```
+
+This is totally valid!  We don't need to worry about the difference between `def` and `val` anymore, because `IO` is referentially transparent.  So you use `def` when you need parameters, and you use `val` when you don't, and you don't have to think about evaluation semantics.  No more subtle bugs caused by accidentally memoizing your effects!
+
+Of course, if `program` is referentially transparent, then clearly repeated values of `program` cannot possibly run the effects it represents multiple times.  For example:
+
+```scala
+program
+program
+program
+
+// must be the same as!
+
+program
+```
+
+If this weren't the case, then we would be in trouble when trying to construct examples like the Haskell one from earlier.  But there is an implication here that is quite profound: `IO` cannot eagerly evaluate its effects, and similarly cannot memoize its results!  If `IO` were to eagerly evaluate or to memoize, then we could no longer replace references to the expression with the expression itself, since that would result in a *different* `IO` instance to be evaluated separately.
+
+This is precisely why `scala.concurrent.Future` is *not* a suitable type for encapsulating effects in this way: it isn't referentially transparent!  `Future` evaluates eagerly (sort of, see below) and memoizes its results, meaning that a `println` inside of a given `Future` will only evaluate *once*, even if the `Future` is sequenced multiple times.  This in turn means that `val x = Future(...); x; x` is not the same as `Future(...); Future(...)`, which is the very definition of a violation of referential transparency.
```

Comment:
> This is precisely why `scala.concurrent.Future` is *not* a suitable type for encapsulating effects in this way: it isn't referentially transparent!

I think we need to be more precise here, since Future is a type and RT is a property of terms. And people like to jump on us when we say this kind of thing. Maybe something like

> This is precisely why `scala.concurrent.Future` is *not* a suitable type for encapsulating effects in this way: constructing a `Future` that will [eventually] side-effect is itself a side-effect!

I would also suggest `val x = Future(...); f(x, x)` as the example because `x; x` uses statements, which don't really have a meaning in FP.